src/lib/core/datetime.c

```diff
@@ -143,3 +143,149 @@ datetime_compare(const struct datetime *lhs, const struct datetime *rhs)
 
 	return COMPARE_RESULT(lhs->nsec, rhs->nsec);
 }
+
+/**
+ * Interval support functions: stringization and operations
+ */
+
+#define NANOS_PER_SEC 1000000000
+
+static inline int64_t
+interval_seconds(const struct datetime_interval *ival)
+{
+	return (int64_t)ival->sec + ival->nsec / NANOS_PER_SEC;
+}
+
+static inline int
+interval_days(const struct datetime_interval *ival)
+{
+	return interval_seconds(ival) / SECS_PER_DAY;
+}
+
+static inline int
+interval_hours(const struct datetime_interval *ival)
+{
+	return interval_seconds(ival) / (60 * 60);
+}
+
+static inline int
+interval_minutes(const struct datetime_interval *ival)
+{
+	return interval_seconds(ival) / 60;
+}
+
+/**
+ * if |nsec| is larger than allowed range 1e9 then modify passed
+ * sec accordingly
+ */
+static void
+denormalize_interval_nsec(int64_t *psec, int *pnsec)
+{
+	assert(psec != NULL);
+	assert(pnsec != NULL);
+	int64_t sec = *psec;
+	int nsec = *pnsec;
+	/*
+	 * nothing to change:
+	 * - if there is small nsec with 0 in sec
+	 * - or if both sec and nsec have the same sign, and nsec is
+	 *   small enough
+	 */
+	if (nsec == 0)
+		return;
+
+	bool zero_or_same_sign = sec == 0 || (sec < 0) == (nsec < 0);
+	if (zero_or_same_sign && labs(nsec) < NANOS_PER_SEC)
+		return;
+
+	sec += nsec / NANOS_PER_SEC;
+	nsec = nsec % NANOS_PER_SEC;
+
+	*psec = sec;
+	*pnsec = nsec;
+}
+
+static size_t
+seconds_str(char *buf, ssize_t len, bool need_sign, long sec, int nsec)
+{
+	sec %= 60;
+	denormalize_interval_nsec(&sec, &nsec);
+	size_t sz = 0;
+	bool is_neg = sec < 0 || (sec == 0 && nsec < 0);
+
+	SNPRINT(sz, snprintf, buf, len, is_neg ? "-" : (need_sign ? "+" : ""));
+	sec = labs(sec);
+	if (nsec != 0) {
+		SNPRINT(sz, snprintf, buf, len, "%ld.%09ld seconds",
+			sec, labs(nsec));
+	} else {
+		SNPRINT(sz, snprintf, buf, len, "%ld seconds", sec);
+	}
+	return sz;
+}
+
+#define SPACE() \
+	if (sz > 0) { \
+		SNPRINT(sz, snprintf, buf, len, ", "); \
+	}
```

Comment:
(If we pass all necessary context to the local macro then it would become as mouthful as `SNPRINT`. But it's local macro, defined few lines above, and used only in a single function to improve complexity, so I believe it's ok to omit arguments and rely upon some implicit assumptions). 

In the last patch, which was introducing "mons intervals" data structure, I've used SPACE multiple times, so now it's becoming more obvious why we need to use it for better readability. Like

```c
	size_t sz = 0;
	if (ival->year != 0) {
		SNPRINT(sz, snprintf, buf, len, "%+d years", ival->year);
		need_sign = false;
	}
	if (ival->month != 0) {
		SPACE();
		SNPRINT(sz, snprintf, buf, len, signed_fmt[need_sign],
			ival->month);
		SNPRINT(sz, snprintf, buf, len, " months");
		need_sign = false;
	}
	if (ival->week != 0) {
		SPACE();
		SNPRINT(sz, snprintf, buf, len, signed_fmt[need_sign],
			ival->week);
		SNPRINT(sz, snprintf, buf, len, " weeks");
		need_sign = false;
	}
	if (ival->day != 0) {
		SPACE();
		SNPRINT(sz, snprintf, buf, len, signed_fmt[need_sign],
			ival->day);
		SNPRINT(sz, snprintf, buf, len, " days");
		need_sign = false;
	}
// and so on...
```